build errors with NO_ASN_TIME and OPENSSL_EXTRA

I'm using #define NO_ASN_TIME as a temporary work around with time not initialized yet on a embedded system.

I get build errors in ssl.c and asn.c, and have to include <time.h> in user_settings.h to get it to compile

ssl.c(23607,17): error C2079: 'ts' uses undefined struct 'tm'
ssl.c(23622,23): warning C4013: 'XTIME' undefined; assuming extern returning int
ssl.c(23629,5): error C2027: use of undefined type 'tm'
\wolfSSL\wolfssl\wolfcrypt\asn.h(1069): message : see declaration of 'tm'
ssl.c(23635,12): error C2224: left of '.tm_year' must have struct/union type
ssl.c(23636,12): error C2224: left of '.tm_year' must have struct/union type
ssl.c(23637,16): error C2224: left of '.tm_year' must have struct/union type
ssl.c(23638,16): error C2224: left of '.tm_year' must have struct/union type
ssl.c(23644,12): error C2224: left of '.tm_year' must have struct/union type
ssl.c(23645,12): error C2224: left of '.tm_year' must have struct/union type
ssl.c(23646,12): error C2224: left of '.tm_year' must have struct/union type
ssl.c(23647,12): error C2224: left of '.tm_year' must have struct/union type
ssl.c(23648,12): error C2224: left of '.tm_year' must have struct/union type
ssl.c(23656,12): error C2224: left of '.tm_mon' must have struct/union type
ssl.c(23657,12): error C2224: left of '.tm_mon' must have struct/union type
ssl.c(23658,12): error C2224: left of '.tm_mday' must have struct/union type
ssl.c(23659,12): error C2224: left of '.tm_mday' must have struct/union type
ssl.c(23660,12): error C2224: left of '.tm_hour' must have struct/union type
ssl.c(23661,12): error C2224: left of '.tm_hour' must have struct/union type
ssl.c(23662,12): error C2224: left of '.tm_min' must have struct/union type
ssl.c(23663,12): error C2224: left of '.tm_min' must have struct/union type
ssl.c(23664,12): error C2224: left of '.tm_sec' must have struct/union type
ssl.c(23665,12): error C2224: left of '.tm_sec' must have struct/union type
ssl.c(23668,21): warning C4013: 'XGMTIME' undefined; assuming extern returning int
ssl.c(23668,35): warning C4047: '=': 'tm *' differs in levels of indirection from 'int'

--- a/src/internal.c
+++ b/src/internal.c
@@ -7309,8 +7309,7 @@ ProtocolVersion MakeDTLSv1_2(void)
 #endif /* WOLFSSL_DTLS */
 
 
-
-
+#ifndef NO_ASN_TIME
 #if defined(USER_TICKS)
 #if 0
     word32 LowResTimer(void)
@@ -7525,6 +7524,7 @@ ProtocolVersion MakeDTLSv1_2(void)
         return (word32)XTIME(0);
     }
 #endif
+#endif /* !NO_ASN_TIME */
 #if !defined(WOLFSSL_NO_CLIENT_AUTH) && \
                ((defined(HAVE_ED25519) && !defined(NO_ED25519_CLIENT_AUTH)) || \
                 (defined(HAVE_ED448) && !defined(NO_ED448_CLIENT_AUTH)))

--- a/src/ssl.c
+++ b/src/ssl.c
@@ -23539,6 +23539,7 @@ int wolfSSL_X509_cmp(const WOLFSSL_X509 *a, const WOLFSSL_X509 *b)
             }
         }
 
+    #ifndef NO_ASN_TIME
         /* print validity */
         {
             char tmp[80];
@@ -23547,6 +23548,7 @@ int wolfSSL_X509_cmp(const WOLFSSL_X509 *a, const WOLFSSL_X509 *b)
                           (int)XSTRLEN("        Validity\n")) <= 0) {
                 return WOLFSSL_FAILURE;
             }
+
             if (wolfSSL_BIO_write(bio, "            Not Before: ",
                           (int)XSTRLEN("            Not Before: ")) <= 0) {
                 return WOLFSSL_FAILURE;
@@ -23591,6 +23593,7 @@ int wolfSSL_X509_cmp(const WOLFSSL_X509 *a, const WOLFSSL_X509 *b)
                 return WOLFSSL_FAILURE;
             }
         }
+    #endif
 
         /* print subject */
         {
@@ -26450,6 +26453,7 @@ void wolfSSL_X509_OBJECT_free_contents(WOLFSSL_X509_OBJECT* obj)
 }
 #endif
 
+#ifndef NO_ASN_TIME
 int wolfSSL_X509_cmp_current_time(const WOLFSSL_ASN1_TIME* asnTime)
 {
     return wolfSSL_X509_cmp_time(asnTime, NULL);
@@ -26534,6 +26538,7 @@ int wolfSSL_X509_cmp_time(const WOLFSSL_ASN1_TIME* asnTime, time_t* cmpTime)
 
     return ret;
 }
+#endif /* !NO_ASN_TIME */
 
 #if defined(OPENSSL_EXTRA) && !defined(NO_ASN_TIME) && !defined(USER_TIME) && \
     !defined(TIME_OVERRIDES)
@@ -26754,6 +26759,7 @@ WOLFSSL_ASN1_INTEGER* wolfSSL_X509_get_serialNumber(WOLFSSL_X509* x509)
 
 #if defined(WOLFSSL_MYSQL_COMPATIBLE) || defined(WOLFSSL_NGINX) || \
     defined(WOLFSSL_HAPROXY) || defined(OPENSSL_EXTRA) || defined(OPENSSL_ALL)
+#ifndef NO_ASN_TIME
 int wolfSSL_ASN1_TIME_print(WOLFSSL_BIO* bio, const WOLFSSL_ASN1_TIME* asnTime)
 {
     char buf[MAX_TIME_STRING_SZ];
@@ -26781,7 +26787,6 @@ int wolfSSL_ASN1_TIME_print(WOLFSSL_BIO* bio, const WOLFSSL_ASN1_TIME* asnTime)
     return ret;
 }
 
-
 char* wolfSSL_ASN1_TIME_to_string(WOLFSSL_ASN1_TIME* t, char* buf, int len)
 {
     WOLFSSL_ENTER("wolfSSL_ASN1_TIME_to_string");
@@ -26802,6 +26807,7 @@ char* wolfSSL_ASN1_TIME_to_string(WOLFSSL_ASN1_TIME* t, char* buf, int len)
 
     return buf;
 }
+#endif /* !NO_ASN_TIME */
 #endif /* WOLFSSL_MYSQL_COMPATIBLE || WOLFSSL_NGINX || WOLFSSL_HAPROXY ||
     OPENSSL_EXTRA*/
 
@@ -29454,6 +29460,7 @@ int wolfSSL_BIO_dump(WOLFSSL_BIO *bio, const char *buf, int length)
     return ret;
 }
 
+#ifndef NO_ASN_TIME
 int wolfSSL_ASN1_UTCTIME_print(WOLFSSL_BIO* bio, const WOLFSSL_ASN1_UTCTIME* a)
 {
     WOLFSSL_ENTER("ASN1_UTCTIME_print");
@@ -29467,7 +29474,6 @@ int wolfSSL_ASN1_UTCTIME_print(WOLFSSL_BIO* bio, const WOLFSSL_ASN1_UTCTIME* a)
 
     return wolfSSL_ASN1_TIME_print(bio, a);
 }
-
 
 /* Checks the ASN1 syntax of "a"
  * returns WOLFSSL_SUCCESS (1)  if correct otherwise WOLFSSL_FAILURE (0) */
@@ -29484,6 +29490,7 @@ int wolfSSL_ASN1_TIME_check(const WOLFSSL_ASN1_TIME* a)
     }
     return WOLFSSL_SUCCESS;
 }
+#endif /* !NO_ASN_TIME */
 
 #ifndef NO_WOLFSSL_STUB
 int wolfSSL_ASN1_TIME_diff(int *pday, int *psec,

--- a/tests/api.c
+++ b/tests/api.c
@@ -20388,7 +20388,7 @@ static void test_wolfSSL_ASN1_TIME_print(void)
 
 static void test_wolfSSL_ASN1_UTCTIME_print(void)
 {
-    #if defined(OPENSSL_EXTRA)
+#if defined(OPENSSL_EXTRA) && !defined(NO_ASN_TIME)
     BIO*  bio;
     ASN1_UTCTIME* utc = NULL;
     unsigned char buf[25];
@@ -20429,7 +20429,7 @@ static void test_wolfSSL_ASN1_UTCTIME_print(void)
     BIO_free(bio);
 
     printf(resultFmt, passed);
-    #endif /* OPENSSL_EXTRA */
+#endif /* OPENSSL_EXTRA && !NO_ASN_TIME */
 }
 
 
@@ -23365,7 +23365,7 @@ static void test_wolfSSL_DES_ecb_encrypt(void)
 
 static void test_wolfSSL_ASN1_TIME_adj(void)
 {
-#if defined(OPENSSL_EXTRA) && !defined(NO_ASN1_TIME) \
+#if defined(OPENSSL_EXTRA) && !defined(NO_ASN_TIME) \
 && !defined(USER_TIME) && !defined(TIME_OVERRIDES)
 
     const int year = 365*24*60*60;
@@ -23455,7 +23455,7 @@ static void test_wolfSSL_ASN1_TIME_adj(void)
 
 static void test_wolfSSL_X509_cmp_time(void)
 {
-#if defined(OPENSSL_EXTRA) && !defined(NO_ASN1_TIME) \
+#if defined(OPENSSL_EXTRA) && !defined(NO_ASN_TIME) \
 && !defined(USER_TIME) && !defined(TIME_OVERRIDES)
     WOLFSSL_ASN1_TIME asn_time;
     time_t t;
@@ -23477,9 +23477,10 @@ static void test_wolfSSL_X509_cmp_time(void)
 
 static void test_wolfSSL_X509_time_adj(void)
 {
-#if defined(OPENSSL_EXTRA) && !defined(NO_ASN1_TIME) && \
+#if defined(OPENSSL_EXTRA) && !defined(NO_ASN_TIME) && \
     !defined(USER_TIME) && !defined(TIME_OVERRIDES) && \
-    defined(USE_CERT_BUFFERS_2048) && !defined(NO_RSA)
+    defined(USE_CERT_BUFFERS_2048) && !defined(NO_RSA) && \
+    !defined(NO_ASN_TIME)
     X509*  x509;
     time_t t, not_before, not_after;
 
@@ -24647,7 +24648,7 @@ static void test_wolfSSL_X509_set_name(void)
 static void test_wolfSSL_X509_set_notAfter(void)
 {
 #if (defined(OPENSSL_ALL) || defined(WOLFSSL_APACHE_HTTPD)) \
-    && !defined(NO_ASN1_TIME) && !defined(USER_TIME) && \
+    && !defined(NO_ASN_TIME) && !defined(USER_TIME) && \
     !defined(TIME_OVERRIDES) && !defined(NO_CERTS) && \
     defined(WOLFSSL_CERT_GEN) && defined(WOLFSSL_CERT_REQ) &&\
     !defined(TIME_T_NOT_64BIT) && !defined(NO_64BIT)
@@ -24702,7 +24703,7 @@ static void test_wolfSSL_X509_set_notAfter(void)
 static void test_wolfSSL_X509_set_notBefore(void)
 {
 #if (defined(OPENSSL_ALL) || defined(WOLFSSL_APACHE_HTTPD)) \
-    && !defined(NO_ASN1_TIME) && !defined(USER_TIME) && \
+    && !defined(NO_ASN_TIME) && !defined(USER_TIME) && \
     !defined(TIME_OVERRIDES) && !defined(NO_CERTS) && \
     defined(WOLFSSL_CERT_GEN) && defined(WOLFSSL_CERT_REQ)
 
@@ -26462,7 +26463,7 @@ static void test_wolfSSL_ASN1_STRING_print_ex(void){
 }
 
 static void test_wolfSSL_ASN1_TIME_to_generalizedtime(void){
-#if defined(OPENSSL_EXTRA) && !defined(NO_ASN1_TIME)
+#if defined(OPENSSL_EXTRA) && !defined(NO_ASN_TIME)
     WOLFSSL_ASN1_TIME *t;
     WOLFSSL_ASN1_TIME *out;
     WOLFSSL_ASN1_TIME *gtime;

--- a/wolfssl/wolfcrypt/settings.h
+++ b/wolfssl/wolfcrypt/settings.h
@@ -2168,6 +2168,12 @@ extern void uITRON4_free(void *p) ;
     #undef WOLFSSL_SMALL_STACK
 #endif
 
+/* The client session cache requires time for timeout */
+#if defined(NO_ASN_TIME) && !defined(NO_SESSION_CACHE)
+    #define NO_SESSION_CACHE
+#endif
+
+
 #ifdef __cplusplus
     }   /* extern "C" */
 #endif

